lib/crypto/src/field/fp.rs

```diff
@@ -0,0 +1,999 @@
+//! This module contains the implementation of a prime field element [`Fp`],
+//! altogether with exact implementations [`Fp64`] for 64-bit, [`Fp128`] for
+//! 128-bit elements and so on.
+//!
+//! Finite field element [`Fp`] wraps a biginteger element in [motgomery form],
+//! which is used for efficient multiplication and division.
+//!
+//! Note that implementation of `Ord` for [`Fp`] compares field elements viewing
+//! them as integers in the range `0, 1, ..., P::MODULUS - 1`.
+//! However, other implementations of `PrimeField` might choose a different
+//! ordering, and as such, users should use this `Ord` for applications where
+//! any ordering suffices (like in a `BTreeMap`), and not in applications
+//! where a particular ordering is required.
+//!
+//! [motgomery form]: https://en.wikipedia.org/wiki/Montgomery_modular_multiplication
+use alloc::string::ToString;
+use core::{
+    cmp::Ordering,
+    fmt::{Debug, Display, Formatter},
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+};
+
+use crypto_bigint::{
+    modular::{
+        constant_mod::{Residue, ResidueParams},
+        montgomery_reduction,
+    },
+    Limb, Uint, Word,
+};
+use educe::Educe;
+use num_traits::{One, Zero};
+
+use crate::field::{group::AdditiveGroup, prime::PrimeField, Field};
+
+/// A trait that specifies the configuration of a prime field.
+/// Also specifies how to perform arithmetic on field elements.
+pub trait FpParams<const LIMBS: usize>: Send + Sync + 'static + Sized {
+    /// The modulus of the field.
+    const MODULUS: Uint<LIMBS>;
+
+    /// A multiplicative generator of the field.
+    /// [`Self::GENERATOR`] is an element having multiplicative order
+    /// `MODULUS - 1`.
+    const GENERATOR: Fp<Self, LIMBS>;
+
+    /// Set `a += b`.
+    fn add_assign(a: &mut Fp<Self, LIMBS>, b: &Fp<Self, LIMBS>) {
```

Comment:
I believe you should use regular Rust traits for math operations.

Reply:
We already have rust math operations for `Fp` types. These ones in `FpParams` just let you to override an algorithm for addition, same as in arkworks library. 
But I'm leaning towards: we don't need math operations in `FpParams` trait at this early stage. And keep it more simple..